src/index.ts

```diff
@@ -142,35 +142,36 @@ export function either<T>(parser: Parser<T>): Parser<{isError: boolean, value: T
   });
 };
 
-// coroutine :: (() -> Iterator (Parser e a s)) -> Parser e a s
-export function coroutine<T>(g: FnReturingParserIterator<T>): Parser<T> {
+// coroutine :: (() -> e) -> Parser e a s
+export function coroutine<T>(scanner: (tokenizer:<K>(parser:Parser<K>)=>K)=>T): Parser<T> {
   return new Parser(function coroutine$state(state) {
-    const generator = g();
-
-    let nextValue = undefined;
-    let nextState = state;
-
-    while (true) {
-      const result = generator.next(nextValue);
-      const value = result.value;
-      const done = result.done;
-
-      if (done) {
-        return updateResult(nextState, value);
-      }
+    let currentValue;
+    let currentState = state;
 
+    const tokenizer = <T>(parser: Parser<T>) => {
+      const value = parser;
       if (!(value && value instanceof Parser)) {
         throw new Error(
-          `[coroutine] yielded values must be Parsers, got ${result.value}.`,
+          `[coroutine] yielded values must be Parsers, got ${value}.`,
         );
       }
-
-      nextState = value.p(nextState);
-      if (nextState.isError) {
-        return nextState;
+      currentState = value.p(currentState);
```

Comment:
This could cause problem. Instead of assigning to currentState, make a new temporary state variable. Then only assign to currentState if there was no error.
This will allow the user to have a try/catch block in the function they pass to coroutine, and then they can choose to handle the error themselves (this is quite a nice improvement over the previous implementation!). If you assign to currentState already, most of the parsers will not run because of the error check they have internally.
Of course, if the user does not have their own try/catch, it will be caught be below.

Reply:
Fixed thia issue